Framework/Muon/src/MuonPairingAsymmetry.cpp

```diff
@@ -0,0 +1,481 @@
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(MuonPairingAsymmetry)
+
+void MuonPairingAsymmetry::init() {
+  std::string emptyString("");
+  std::vector<int> defaultGrouping1 = {1};
+  std::vector<int> defaultGrouping2 = {2};
+
+  declareProperty(
+      Mantid::Kernel::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+          "OutputWorkspace", emptyString, Direction::Output),
+      "The workspace which will hold the results of the asymmetry "
+      "calculation.");
+
+  declareProperty("PairName", emptyString,
+                  "The name of the pair. Must "
+                  "contain at least one alphanumeric "
+                  "character.",
+                  Direction::Input);
+
+  declareProperty(
+      "Alpha", 1.0, boost::make_shared<MandatoryValidator<double>>(),
+      "Alpha parameter used in the asymmetry calculation.", Direction::Input);
+
+  declareProperty("SpecifyGroupsManually", false,
+                  "Specify the pair of groups manually");
+
+  // Select groups via workspaces
+
+  declareProperty(Mantid::Kernel::make_unique<WorkspaceProperty<Workspace>>(
+                      "InputWorkspace1", emptyString, Direction::Input,
+                      PropertyMode::Optional),
+                  "Input workspace containing data from grouped detectors.");
+
+  declareProperty(Mantid::Kernel::make_unique<WorkspaceProperty<Workspace>>(
+                      "InputWorkspace2", emptyString, Direction::Input,
+                      PropertyMode::Optional),
+                  "Input workspace containing data from grouped detectors.");
+
+  setPropertySettings("InputWorkspace1",
+                      make_unique<Kernel::EnabledWhenProperty>(
+                          "SpecifyGroupsManually", Kernel::IS_EQUAL_TO, "0"));
+  setPropertySettings("InputWorkspace2",
+                      make_unique<Kernel::EnabledWhenProperty>(
+                          "SpecifyGroupsManually", Kernel::IS_EQUAL_TO, "0"));
+
+  // Specify groups manually
+
+  declareProperty(
+      Mantid::Kernel::make_unique<WorkspaceProperty<WorkspaceGroup>>(
+          "InputWorkspace", emptyString, Direction::Input,
+          PropertyMode::Optional),
+      "Input workspace containing data from detectors which are to "
+      "be grouped.");
+  setPropertySettings("InputWorkspace",
+                      make_unique<Kernel::EnabledWhenProperty>(
+                          "SpecifyGroupsManually", Kernel::IS_EQUAL_TO, "1"));
+
+  declareProperty(make_unique<ArrayProperty<int>>(
+                      "Group1", defaultGrouping1,
+                      IValidator_sptr(new NullValidator), Direction::Input),
+                  "The grouping of detectors, comma separated list of detector "
+                  "IDs or hyphenated ranges of IDs.");
+  declareProperty(make_unique<ArrayProperty<int>>(
+                      "Group2", defaultGrouping2,
+                      IValidator_sptr(new NullValidator), Direction::Input),
```

Comment:
Why not share the `NullValidator`s?

Reply:
I'm not quite sure why but sharing the NullValidators seems to lead to a segfault when you start up mantid.

Reply:
Wait a second I'm being a muppet just need to cast to another const.

Reply:
I see. Maybe it is a bug? Why is `NullValidator` even used here? I would have thought it is the default anyways. It is not a big issue, however, I think we can leave this as is.